vm_eval.c

```diff
@@ -1642,10 +1642,118 @@ get_eval_default_path(void)
+        for (int local_index = 0; local_index < locals_count; local_index++) {
+            pm_string_t *scope_local = &options_scope->locals[local_index];
+            const char *name = rb_id2name(ISEQ_BODY(iseq)->local_table[local_index]);
+            if (name) pm_string_constant_init(scope_local, name, strlen(name));
+        }
+
+        iseq = ISEQ_BODY(iseq)->parent_iseq;
+    }
+
+    // Add our empty local scope at the very end of the array for our eval
+    // scope's locals.
+    pm_options_scope_init(&result.options.scopes[scopes_count], 0);
+    VALUE error = pm_parse_string(&result, src, fname);
+
+    // If the parse failed, clean up and raise.
+    if (error != Qnil) {
+        pm_parse_result_free(&result);
+        rb_exc_raise(error);
+    }
+    // Create one scope node for each scope passed in, initialize the local
+    // lookup table with all the local variable information attached to the
+    // scope used by the parser.
+    pm_scope_node_t *node = &result.node;
+    iseq = parent;
+
+    for (int scopes_index = 0; scopes_index < scopes_count; scopes_index++) {
+        pm_scope_node_t *parent_scope = ruby_xcalloc(1, sizeof(pm_scope_node_t));
+        if (parent_scope == NULL) abort();
+
+        pm_options_scope_t *options_scope = &result.options.scopes[scopes_count - scopes_index - 1];
+        parent_scope->parser = &result.parser;
+        parent_scope->index_lookup_table = st_init_numtable();
+
+        int locals_count = ISEQ_BODY(iseq)->local_table_size;
+        parent_scope->local_table_for_iseq_size = locals_count;
+        pm_constant_id_list_init(&parent_scope->locals);
+
+        for (int local_index = 0; local_index < locals_count; local_index++) {
+            const pm_string_t *scope_local = &options_scope->locals[local_index];
+
+            pm_constant_id_t constant_id = 0;
+            if (pm_string_length(scope_local) > 0) {
+                constant_id = pm_constant_pool_insert_constant(
+                        &result.parser.constant_pool, pm_string_source(scope_local),
+                        pm_string_length(scope_local));
+                st_insert(parent_scope->index_lookup_table, (st_data_t)constant_id, (st_data_t)local_index);
+            }
+            pm_constant_id_list_append(&parent_scope->locals, constant_id);
+        }
+
+        node->previous = parent_scope;
+        node = parent_scope;
+        iseq = ISEQ_BODY(iseq)->parent_iseq;
+    }
+
+    iseq = pm_iseq_new_eval(&result.node, name, fname, fname, line, parent, 0);
+
+    pm_scope_node_t *prev = result.node.previous;
+    while (prev) {
+        pm_scope_node_t *next = prev->previous;
+        ruby_xfree(prev);
+        prev = next;
+    }
+    pm_parse_result_free(&result);
+
+    return iseq;
+}
+
 static const rb_iseq_t *
 eval_make_iseq(VALUE src, VALUE fname, int line,
                const struct rb_block *base_block)
 {
+    if (*rb_ruby_prism_ptr()) {
+        return pm_eval_make_iseq(src, fname, line, base_block);
```

Comment:
It looks like we might need a couple of things lower in this function as well. Mostly:

```c
    if (!fname) {
        fname = rb_source_location(&line);
    }
    if (!UNDEF_P(fname)) {
        if (!NIL_P(fname)) fname = rb_fstring(fname);
    }
    else {
        fname = get_eval_default_path();
    }
```

and whatever `isolated_depth` is

and